File: solutions/typescript/079-gas-station.ts
/**
 * Problem 79: Gas Station (LeetCode 134)
 * Difficulty: Med
 * Language: TypeScript
 */
function canCompleteCircuit(gas: number[], cost: number[]): number {
  const diffs = gas.map((g, i) => g - cost[i]);
  const result = diffs.reduce(
    (acc, d, i) => {
      const total = acc.total + d;
      const tank = acc.tank + d;
      return tank < 0
        ? { total, tank: 0, start: i + 1 }
        : { total, tank, start: acc.start };
    },
    { total: 0, tank: 0, start: 0 }
  );
  return result.total >= 0 ? result.start : -1;
}

File: solutions/rust/079-gas-station.rs
/**
 * Problem 79: Gas Station (LeetCode 134)
 * Difficulty: Med
 * Language: Rust
 */
fn can_complete_circuit(gas: Vec<i32>, cost: Vec<i32>) -> i32 {
    let (total, _, start) = gas.iter().zip(cost.iter()).enumerate()
        .fold((0, 0, 0), |(tot, tank, s), (i, (g, c))| {
            let d = g - c;
            let (tot2, tank2) = (tot + d, tank + d);
            if tank2 < 0 { (tot2, 0, i as i32 + 1) } else { (tot2, tank2, s) }
        });
    if total >= 0 { start } else { -1 }
}
